feat: implement grc1155

Jammyaa opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!--- If you need to use a more detailed template, append the query param template=detailed_pr_template.md to the URL -->

# Description
- Implement GRC1155 base methods
- Implement foo1155 realm based on GRC1155
- Add tests for p/demo/grc/grc1155 and r/demo/foo1155

# How has this been tested?
done all cases for p/demo/grc/grc1155 and r/demo/foo1155

thehowl commented on examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token.gno:
```diff
@@ -0,0 +1,443 @@
+	s.balances.Set(fromBalanceKey, fromBalance)
+	s.balances.Set(toBalanceKey, toBalance)
+
+	event := TransferSingleEvent{caller, from, to, tid, amount}
+	emit(&event)
```
since we're not reusing event, `emit(&TransferSingleEvent{caller, from, to, tid, amount})`? (here as well as the occurrences below)

thehowl commented on examples/gno.land/p/demo/grc/grc1155/util.gno:
```diff
@@ -0,0 +1,18 @@
+
+var zeroAddress = std.Address("")
+
+func isValidAddress(addr std.Address) error {
+	if addr.String() == "" {
+		return ErrInvalidAddress
+	}
+	return nil
+}
```
maybe change this to return a bool instead of an error? this way, the callers can do the following

```go
if !isValidAddress(from) || !isValidAddress(to) {
    return ErrInvalidAddress
}
```

in case we should distinguish the errors we can always change this as this is unexported.

thehowl commented on examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token.gno:
```diff
@@ -0,0 +1,443 @@
+		return ErrInvalidAddress
+	}
+	if err := isValidAddress(to); err != nil {
+		return ErrInvalidAddress
+	}
+	if from == to {
+		return ErrCannotTransferToSelf
+	}
+
+	caller := std.GetOrigCaller()
+	batch := []TokenID{tid}
+	amounts := []uint64{amount}
+	s.beforeTokenTransfer(caller, from, to, batch, amounts)
+
+	fromBalance, err := s.BalanceOf(from, tid)
+	if err != nil {
+		return err
+	}
+	if fromBalance < amount {
+		return ErrInsufficientBalance
+	}
+	toBalance, err := s.BalanceOf(to, tid)
+	if err != nil {
+		return err
+	}
+
+	fromBalance -= amount
+	toBalance += amount
+	fromBalanceKey := string(tid) + ":" + from.String()
+	toBalanceKey := string(tid) + ":" + to.String()
+	s.balances.Set(fromBalanceKey, fromBalance)
+	s.balances.Set(toBalanceKey, toBalance)
+
+	event := TransferSingleEvent{caller, from, to, tid, amount}
+	emit(&event)
+
+	s.afterTokenTransfer(caller, from, to, batch, amounts)
+
+	return nil
+}
```
what if, to avoid duplicating logic, this entire function was just:

```go
// called from SafeTransferFrom
s.safeBatchTransferFrom(from, to, []TokenID{tid}, []uint64{amount})
```

and the event emission was pulled out of `safeTransferFrom` and batch version, and were put instead in their exported counterparts? (so each function actually emits the correct event)

thehowl commented on examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token.gno:
```diff
@@ -0,0 +1,443 @@
+}
+
+// Helper for SafeMint()
+func (s *basicGRC1155Token) mint(to std.Address, tid TokenID, amount uint64) error {
+	if err := isValidAddress(to); err != nil {
+		return err
+	}
+
+	caller := std.GetOrigCaller()
+	batch := []TokenID{tid}
+	amounts := []uint64{amount}
+	s.beforeTokenTransfer(caller, zeroAddress, to, batch, amounts)
+
+	toBalance, err := s.BalanceOf(to, tid)
+	if err != nil {
+		return err
+	}
+	toBalance += amount
+	toBalanceKey := string(tid) + ":" + to.String()
+	s.balances.Set(toBalanceKey, toBalance)
+
+	event := TransferSingleEvent{caller, zeroAddress, to, tid, amount}
+	emit(&event)
+
+	s.afterTokenTransfer(caller, zeroAddress, to, batch, amounts)
+
+	return nil
+}
```
again, I think for reducing code and logic this function could be removed and the `Mint` method could instead just call `mintBatch` with the arrays, making sure that the events are emitted accordingly.

thehowl commented on examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token.gno:
```diff
@@ -0,0 +1,443 @@
+	return nil
+}
+
+// Helper for Burn()
+func (s *basicGRC1155Token) burn(from std.Address, tid TokenID, amount uint64) error {
+	if err := isValidAddress(from); err != nil {
+		return err
+	}
+
+	caller := std.GetOrigCaller()
+	batch := []TokenID{tid}
+	amounts := []uint64{amount}
+	s.beforeTokenTransfer(caller, from, zeroAddress, batch, amounts)
+
+	fromBalance, err := s.BalanceOf(from, tid)
+	if err != nil {
+		return err
+	}
+	if fromBalance < amount {
+		return ErrBurnAmountExceedsBalance
+	}
+	fromBalance -= amount
+	fromBalanceKey := string(tid) + ":" + from.String()
+	s.balances.Set(fromBalanceKey, fromBalance)
+
+	event := TransferSingleEvent{caller, from, zeroAddress, tid, amount}
+	emit(&event)
+
+	s.afterTokenTransfer(caller, from, zeroAddress, batch, amounts)
+
+	return nil
+}
```
same for `Burn` and `burnBatch`

thehowl commented on examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token.gno:
```diff
@@ -0,0 +1,443 @@
+}
+
+func (s *basicGRC1155Token) setUri(newUri string) {
+	s.uri = newUri
```
Should we not emit `UpdateURIEvent`?

thehowl commented on examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token_test.gno:
```diff
@@ -0,0 +1,390 @@
+	"gno.land/r/demo/users"
+)
+
+var dummyURI = "ipfs://xyz"
```
const?

thehowl commented on examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token_test.gno:
```diff
@@ -0,0 +1,390 @@
+	if balanceBatch[1] != 140 {
+		t.Errorf("expected: (%d), got: (%d)", 140, balanceBatch[1])
+	}
+}
```
please, in general, add some more tests regarding error conditions. for instance:

* should not be able to get the balance of an invalid address
* should not be able to transfer tokens if my balance is less than what i request
* giving bad inputs to batch operations fails (len(addr) != len(amount)), and giving them nothing is a no-op.

thehowl commented on examples/gno.land/p/demo/grc/grc1155/util.gno:
```diff
@@ -0,0 +1,18 @@
+	"std"
+)
+
+var zeroAddress = std.Address("")
```
since the Address type is just a string, `const zeroAddress std.Address = ""`?

thehowl reviewed (changes requested):
Thank you for your contribution! I've found a few things to change and some places where the code can be greatly simplified and reduced in size, thank you in advance for addressing these issues :)

thehowl commented:
You'll also need to add the package and the realm to `cmd/gnoland/main.go` so that they're loaded on the staging and any testnets. :)

Jammyaa commented:
Ok, I will resolve all issues. Thank you for your quality comment!

thehowl reviewed (commented):
rest looks good :)

Jammyaa commented on examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token.gno:
```diff
@@ -0,0 +1,443 @@
+}
+
+// Returns new basic SFT
+func NewBasicGRC1155Token(uri string) *basicGRC1155Token {
```
Ok, I will consider this suggestion. Thank you.

Jammyaa replied:
Sorry, but if you modify this return type from *basicGRC1155Token by IGRC1155, you will see the panic in your terminal. 

And if you see p/demo/grc721 of @harry-hov, he uses *basicNFT (not IGRC721) in the return type of NewBasicNFT() method, similar to me. 

One more thing, I think that should not expose basicGRC1155Token and basicNFT struct to end user directly, so I use lowercase to make it only be accessed within the grc1155 package.

So what's the problem with using *basicGRC1155Token or *basicNFT? Thank you very much @thehowl

thehowl replied:
Hmm... I've reviewed and tried putting the interface. The tests you've made kind of rely on the unexported methods, so it makes sense to keep things like this at the moment. Ideally, we'd still get into a state where in a testing environment we can manipulate the caller address, and refactor the code that currently returns an unexported struct when they should return an interface after that.

The reasoning why I personally avoid returning unexported structs is that in the context of normal Go, with all the tooling etc., from a godoc page you can only see the methods of the exported types. Of course, tooling still knows how to work with unexported types, so for instance `gopls` and other autocompletion tools will still give you documentation for the methods of an unexported type, but in general while it makes it easier to go look up the implementation code of the type, it makes it harder to go look up the documentation from a godoc page.

I'll create a separate issue for this, though.

thehowl reviewed (approved):
looks good to me, thank you for your time and patience.

Pinging @moul for a last review :)

Jammyaa commented:
Hi sir, I have addressed all issues that Morgan comments, and got his approval, so could you help me review the last time and merge this code? Thank you very much @moul

zivkovicmilos commented on examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token.gno:
```diff
@@ -0,0 +1,347 @@
+	emit(&UpdateURIEvent{newUri})
+}
+
+func (s *basicGRC1155Token) beforeTokenTransfer(operator, from, to std.Address, batch []TokenID, amounts []uint64) {
+	// TODO: Implementation
+}
+
+func (s *basicGRC1155Token) afterTokenTransfer(operator, from, to std.Address, batch []TokenID, amounts []uint64) {
+	// TODO: Implementation
+}
+
+func (s *basicGRC1155Token) doSafeTransferAcceptanceCheck(operator, from, to std.Address, tid TokenID, amount uint64) bool {
+	// TODO: Implementation
+	return true
+}
+
+func (s *basicGRC1155Token) doSafeBatchTransferAcceptanceCheck(operator, from, to std.Address, batch []TokenID, amounts []uint64) bool {
+	// TODO: Implementation
+	return true
+}
```
Are these leftover?

zivkovicmilos replied:
Got it, no problem, just wanted to double check

zivkovicmilos reviewed (approved):
Thank you for the contribution 🙏 

Looks good 💯

Jammyaa commented:
Excuse me, I don't know if there is still a problem with this contribution of mine, it has not been merged after being approved for a long time. 
I am continuing my work by implementing soul bound token for Gnoland, can someone merge and close the PR so that I can continue to contribute from my fork branch. Thank you very much. @moul @thehowl
